runtime/runtime_storage.go

```diff
@@ -211,41 +212,122 @@ func (s *interpreterRuntimeStorage) writeValue(
 //
 func (s *interpreterRuntimeStorage) writeCached() {
 
+	type writeItem struct {
+		storageKey storageKey
+		value      interpreter.Value
+	}
+
+	var items []writeItem
+
 	for fullKey, entry := range s.cache {
 
 		if !entry.mustWrite && entry.value != nil && !entry.value.Modified() {
 			continue
 		}
 
+		items = append(items, writeItem{
+			storageKey: fullKey,
+			value:      entry.value,
+		})
+	}
+
+	// Don't use a for-range loop, as keys are added while iterating
+	for len(items) > 0 {
+		var item writeItem
+		item, items = items[0], items[1:]
+
 		var newData []byte
-		if entry.value != nil {
+		if item.value != nil {
+			var deferrals *interpreter.EncodingDeferrals
 			var err error
-			reportMetric(
-				func() {
-					newData, err = interpreter.EncodeValue(entry.value)
-				},
-				s.runtimeInterface,
-				func(metrics Metrics, duration time.Duration) {
-					metrics.ValueEncoded(duration)
-				},
-			)
+			newData, deferrals, err = s.encodeValue(item.value, item.storageKey.key)
 			if err != nil {
 				panic(err)
 			}
+
+			for deferredKey, deferredValue := range deferrals.Values {
+
+				deferredStorageKey := storageKey{
+					address: item.storageKey.address,
+					key:     deferredKey,
+				}
+
+				if !deferredValue.Modified() {
+					continue
+				}
+
+				items = append(items, writeItem{
+					storageKey: deferredStorageKey,
+					value:      deferredValue,
+				})
+			}
+
+			for _, deferralMove := range deferrals.Moves {
+
+				s.move(
+					deferralMove.DeferredOwner,
+					deferralMove.DeferredStorageKey,
+					deferralMove.NewOwner,
+					deferralMove.NewStorageKey,
+				)
+			}
```

Comment:
So are these loops just to allow you to skip over deferred values that have not changed? Otherwise all deferred writes are just added to `items` and written, right?

Reply:
Yeah, the first loop above is the interaction of optimizing writeback by only writing modified values and the new deferral feature: If a value is deferred and was loaded from storage into memory, then we should only write it back if it was modified.

The moves have a different reason: If a dictionary with deferred values is moved from account to another, the deferred values / storage keys have to be moved over as well